### PR TITLE
Fix (flexml): PTQ calibration fix for training/eval mode

### PR DESCRIPTION
**Pull request by Giuseppe5**
Fix for correct handling of training state of layers and proxies during calibration.

BiasCorrection requires more adaptations to match this interface.

```python
self.model = quantize_flexml(self.model)
self.model.eval()
with torch.no_grad():
    with do_calibration(self.model):
        for i, (images, _) in enumerate(self.dataloader_calibration):
            out = self.model(images)
```

### CODE REVIEW

File: src/brevitas/graph/calibrate.py [hunk 1]
```diff
@@ -43,6 +43,20 @@ def finalize_collect_stats(module):
         module.counter = module.collect_stats_steps
 
 
+class do_calibration(object):
```
Comment (volcacius) [inline src/brevitas/graph/calibrate.py:46]: Rename to calibration_mode

File: src/brevitas/graph/calibrate.py [hunk 2]
```diff
@@ -43,6 +43,20 @@ def finalize_collect_stats(module):
...
+    def __init__(self, model, enable=True):
```
Comment (volcacius) [inline src/brevitas/graph/calibrate.py:47]: Rename to enabled

File: src/brevitas/graph/calibrate.py [hunk 3]
```diff
@@ -43,6 +43,20 @@ def finalize_collect_stats(module):
...
+        self.model = model
+        self.current_state = model.training
+        self.disable_quant_inference = DisableQuantInference()
+        self.enable=enable
+
+    def __enter__(self):
```
Comment (volcacius) [inline src/brevitas/graph/calibrate.py:53]: This way when enabled=False it stills puts the model in training mode. It should be more something like: 
__enter__:
if enabled:
    self.disable_quant_inference.apply(self.model, is_training=True, quantization_enabled=False)

__exit__:
    self.disable_quant_inference.apply(self.model, is_training=current_state, quantization_enabled=True)

File: src/brevitas/graph/calibrate.py [hunk 4]
```diff
@@ -76,42 +90,49 @@ def disable_act_quant_hook(self, module, inp, output):
...
 class DisableQuantInference(DisableEnableQuantization):
 
-    def apply(self, model, *model_args, **model_kwargs):
-        self.disable_act_quantization(model)
-        self.disable_param_quantization(model)
-        output = model(*model_args, **model_kwargs)
-        self.enable_act_quantization(model)
-        self.enable_param_quantization(model)
-        return output
+    def apply(self, model, is_training, enable, *model_args, **model_kwargs):
```
Comment (volcacius) [inline src/brevitas/graph/calibrate.py:129]: rename enable to quantization_enabled

**Review by volcacius — CHANGES_REQUESTED** [review]

**Comment by volcacius** [comment]
The jupyter tvm tutorial needs to be updated too

File: tests/brevitas/graph/test_calibration.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    
+    return scale
+
+@given(inp=float_tensor_random_size_st())
+def test_scale_factors_ptq_calibration_po2(inp):
+    class TestModel(nn.Module):
+
+        def __init__(self):
+            super(TestModel, self).__init__()
+            self.act = qnn.QuantIdentity(Int8ActPerTensorFixedPoint)
```
Comment (volcacius) [inline tests/brevitas/graph/test_calibration.py:31]: act_quant=Int8ActPerTensorFixedPoint, let's not assume it's the first arg

File: src/brevitas/graph/calibrate.py [hunk 6]
```diff
@@ -45,6 +46,21 @@ def finalize_collect_stats(module):
         module.counter = module.collect_stats_steps
 
 
+class calibration_mode(object):
+    def __init__(self, model, enabled=True):
+        self.model = model
+        self.current_state = model.training
```
Comment (volcacius) [inline src/brevitas/graph/calibrate.py:52]: this should be called something like previous_training_state

File: src/brevitas/graph/calibrate.py [hunk 7]
```diff
@@ -85,42 +101,49 @@ def disable_act_quant_hook(self, module, inp, output):
...
 
-    def enable_param_quantization(self, model):
+    def enable_param_quantization(self, model, is_training):
         for module in model.modules():
             if isinstance(module, _PARAM_PROXIES):
                 module.disable_quant = False
+                module.train(is_training)
 
 
 class DisableQuantInference(DisableEnableQuantization):
```
Comment (volcacius) [inline src/brevitas/graph/calibrate.py:138]: We don't need a DisableQuantInference class anymore as inference is performed externally now. The apply method can be moved to DisableEnableQuantization and BiasCorrection can inherit that and then override it.

File: tests/brevitas/graph/test_calibration.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    assert torch.allclose(expected_scale, scale)
+
+
+def test_calibration_training_state():
+    
+    class TestModel(nn.Module):
+
+        def __init__(self):
+            super(TestModel, self).__init__()
+            self.act = qnn.QuantIdentity(Int8ActPerTensorFixedPoint)
```
Comment (volcacius) [inline tests/brevitas/graph/test_calibration.py:55]: act_quant=

File: tests/brevitas/graph/test_calibration.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+        
+    model = TestModel()
+    model.eval()
+    with torch.no_grad():
+        with calibration_mode(model):
+            assert model.act.act_quant.training == True
+            assert model.training == False
+    
+    assert model.act.act_quant.training == False
+    assert model.training == False
```
Comment (volcacius) [inline tests/brevitas/graph/test_calibration.py:68]: Final newline

**Review by volcacius — CHANGES_REQUESTED** [review]

File: src/brevitas/graph/calibrate.py (hunk 7 above)
Comment (Giuseppe5) [reply src/brevitas/graph/calibrate.py:138]: Ok. Maybe in #460 we could move away from the "apply" method and simply use "__call__".

Comment (volcacius) [reply src/brevitas/graph/calibrate.py]: let's do it in a separate PR over all graph transformations, it's more work
